gammapy/analysis/steps.py

```diff
@@ -0,0 +1,498 @@
+            self.analysis.log.debug(
+                f"Creating ReflectedRegionsBackgroundMaker with arguments {bkg_maker_config}"
+            )
+        else:
+            self.analysis.log.warning("No background maker set. Check configuration.")
+        return bkg_maker
+
+    def _map_making(self):
+        """Make maps and datasets for 3d analysis"""
+        datasets_settings = self.analysis.config.datasets
+        offset_max = datasets_settings.geom.selection.offset_max
+
+        self.analysis.log.info("Creating reference dataset and makers.")
+        stacked = self._create_reference_dataset(name="stacked")
+
+        maker = self._create_dataset_maker()
+        maker_safe_mask = self._create_safe_mask_maker()
+        bkg_maker = self._create_background_maker()
+
+        makers = [maker, maker_safe_mask, bkg_maker]
+        makers = [maker for maker in makers if maker is not None]
+
+        self.analysis.log.info("Start the data reduction loop.")
+
+        datasets_maker = DatasetsMaker(makers,
+                                      stack_datasets=datasets_settings.stack,
+                                      n_jobs=self.analysis.config.general.n_jobs,
+                                      cutout_mode='trim',
+                                      cutout_width=2 * offset_max)
+        self.analysis.datasets = datasets_maker.run(stacked, self.analysis.observations)
+        #TODO: move progress bar to DatasetsMaker but how with multiprocessing ?
+
+
+    def _spectrum_extraction(self):
+        """Run all steps for the spectrum extraction."""
+        self.analysis.log.info("Reducing spectrum datasets.")
+        datasets_settings = self.analysis.config.datasets
+        dataset_maker = self._create_dataset_maker()
+        safe_mask_maker = self._create_safe_mask_maker()
+        bkg_maker = self._create_background_maker()
+
+        reference = self._create_reference_dataset()
+
+        datasets = []
+        for obs in progress_bar(self.analysis.observations, desc="Observations"):
+            self.analysis.log.debug(f"Processing observation {obs.obs_id}")
+            dataset = dataset_maker.run(reference.copy(), obs)
+            if bkg_maker is not None:
+                dataset = bkg_maker.run(dataset, obs)
+                if dataset.counts_off is None:
+                    self.analysis.log.debug(
+                        f"No OFF region found for observation {obs.obs_id}. Discarding."
+                    )
+                    continue
+            dataset = safe_mask_maker.run(dataset, obs)
+            self.analysis.log.debug(dataset)
+            datasets.append(dataset)
+        self.analysis.datasets = Datasets(datasets)
+
+        if datasets_settings.stack:
+            stacked = self.analysis.datasets.stack_reduce(name="stacked")
+            self.analysis.datasets = Datasets([stacked])
+
+
+def make_energy_axis(axis, name="energy"):
```

Comment:
I haven't thought this through, but we could maybe even introduce and API like, `MapAxis.from_analysis_config(config=)` and `Maker.from_analysis_config()`...